File: benchmarks/FlutterAppBench/lib/programs/regex.dart
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Isaac Gouy, RegExp from Jos Hirth's program
*/

// 21/05/2022
// https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/regexredux-dartexe-4.html

import 'dart:io';

void runRegex(String s, IOSink writer) {
    final z = s.replaceAll(RegExp('>.*\n|\n', multiLine: true), '');
    printPatternMatches(z, writer);
    writer.write('\n${s.length}\n${z.length}\n${magicLengthOf(z)}');
}

void printPatternMatches(String s, IOSink writer) {
  const simple = [
    'agggtaaa|tttaccct',
    '[cgt]gggtaaa|tttaccc[acg]',
    'a[act]ggtaaa|tttacc[agt]t',
    'ag[act]gtaaa|tttac[agt]ct',
    'agg[act]taaa|ttta[agt]cct',
    'aggg[acg]aaa|ttt[cgt]ccct',
    'agggt[cgt]aa|tt[acg]accct',
    'agggta[cgt]a|t[acg]taccct',
    'agggtaa[cgt]|[acg]ttaccct'
  ];

  for (var each in simple) {
    final regex = RegExp(each, multiLine: true);
    writer.writeln('$each ${regex.allMatches(s).length}');
  }
}

int magicLengthOf(String s) {
  const magic = [
    ['tHa[Nt]', '<4>'],
    ['aND|caN|Ha[DS]|WaS', '<3>'],
    ['a[NSt]|BY', '<2>'],
    ['<[^>]*>', '|'],
    ['\\|[^|][^|]*\\|', '-'],
  ];

  for (var each in magic) {
    final regex = RegExp(each.first, multiLine: true);
    s = s.replaceAll(regex, each.last);
  }
  return s.length;
}

File: benchmarks/RNAppBench/benchmarks/regex.js
import fs from "react-native-fs";
import { write } from "./writeFile";

async function startRegex(fileName) {

  const ret = await fs.readFileAssets('files/' + fileName + '.txt', "utf8");
  await regex(ret);

  return 'finished';
}

async function regex(data) {

  var output;

  await mainThread(data);

  async function mainThread(data) {
    const regExps = [
      /agggtaaa|tttaccct/gi,
      /[cgt]gggtaaa|tttaccc[acg]/gi,
      /a[act]ggtaaa|tttacc[agt]t/gi,
      /ag[act]gtaaa|tttac[agt]ct/gi,
      /agg[act]taaa|ttta[agt]cct/gi,
      /aggg[acg]aaa|ttt[cgt]ccct/gi,
      /agggt[cgt]aa|tt[acg]accct/gi,
      /agggta[cgt]a|t[acg]taccct/gi,
      /agggtaa[cgt]|[acg]ttaccct/gi,
    ];

    const initialLen = data.length;
    data = data.replace(/^>.*\n|\n/gm, '');
    const cleanedLen = data.length;

    for (let j = 0; j < regExps.length; j++) {
      const re = regExps[j];
      const m = data.match(re);
      output += re.source + " " + (m ? m.length : 0) + "\n";
    }

    const endLen = data
      .replace(/tHa[Nt]/g, '<4>')
      .replace(/aND|caN|Ha[DS]|WaS/g, '<3>')
      .replace(/a[NSt]|BY/g, '<2>')
      .replace(/<[^>]*>/g, '|')
      .replace(/\|[^|][^|]*\|/g, '-').length;

    output += `\n${initialLen}\n${cleanedLen}\n${endLen}`;

    await write(output); 

    return 'finished';
  }


}



export default startRegex;

